### PR TITLE
Add test for BCI systemd in podman

### PR DESCRIPTION
**Pull request by ilausuch**
Testing systemd in a container for the podman engine

- Related ticket: https://progress.opensuse.org/issues/134687
- Verification run: http://openqa.suse.de/tests/12139439

### CODE REVIEW

**Comment by mloviska** [comment]
Isn't this already tested in BCI test framework?

File: tests/containers/podman_bci_systemd.pm [hunk 1]
```diff
@@ -0,0 +1,61 @@
+# SUSE's openQA tests
+#
+# Copyright 2023 SUSE LLC
+# SPDX-License-Identifier: FSFAP
+
+# Package: podman
+# Summary: Test podman with systemd
+# Maintainer: Richard Brown <rbrown@suse.com>
```
Comment (grisu48) [inline tests/containers/podman_bci_systemd.pm:8]: Really?

Comment (ilausuch) [reply tests/containers/podman_bci_systemd.pm:8]: haha, copy

Comment (ilausuch) [reply tests/containers/podman_bci_systemd.pm]: Solved

File: tests/containers/podman_bci_systemd.pm [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+# Copyright 2023 SUSE LLC
+# SPDX-License-Identifier: FSFAP
+
+# Package: podman
+# Summary: Test podman with systemd
+# Maintainer: qa-c@suse.de
+
+use Mojo::Base 'containers::basetest';
+use testapi;
+use utils qw(script_retry);
```
Comment (grisu48) [inline tests/containers/podman_bci_systemd.pm:12]: Not used

File: tests/containers/podman_bci_systemd.pm [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+use containers::utils qw(check_min_runtime_version);
```
Comment (grisu48) [inline tests/containers/podman_bci_systemd.pm:13]: Not used

File: tests/containers/podman_bci_systemd.pm [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+use serial_terminal 'select_serial_terminal';
+use version_utils qw(is_sle is_opensuse is_staging);
```
Comment (grisu48) [inline tests/containers/podman_bci_systemd.pm:15]: Not used

File: tests/containers/podman_bci_systemd.pm [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+use containers::k8s qw(install_k3s uninstall_k3s);
```
Comment (grisu48) [inline tests/containers/podman_bci_systemd.pm:16]: Not used

File: tests/containers/podman_bci_systemd.pm [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+use Utils::Architectures qw(is_ppc64le);
```
Comment (grisu48) [inline tests/containers/podman_bci_systemd.pm:17]: Not used

File: tests/containers/podman_bci_systemd.pm [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    record_info('Test', 'Launch a container with systemd');
+    assert_script_run("podman run -d -p 80:80 --name nginx $image");
+
+    record_info('Test', 'Install nginx');
+    assert_script_run("podman exec -t nginx zypper -n in nginx");
+
+    record_info('Test', 'Start nginx');
+    assert_script_run("podman exec -t nginx systemctl start nginx");
+
+    record_info('Nginx service status', script_output("podman exec -t nginx systemctl status nginx"));
```
Comment (grisu48) [inline tests/containers/podman_bci_systemd.pm:37]: I would replace this by a `validate_script_output_retry`.

File: tests/containers/podman_bci_systemd.pm [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+    record_info('Test', 'Curl localhost');
+    assert_script_run("podman exec -t nginx curl http://localhost");
```
Comment (grisu48) [inline tests/containers/podman_bci_systemd.pm:40]: This throws a 403 and `curl` is still happy. Let's also create a predefined page, which we can validate

```suggestion
    assert_script_run("podman exec -t nginx bash -c 'echo testpage123-content > /srv/www/htdocs/index.html'
    validate_script_output("podman exec -t nginx curl -sfL http://localhost", qr/testpage123-content/ );
```

File: tests/containers/podman_bci_systemd.pm [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+    record_info('Test', 'Curl localhost from host');
+    assert_script_run("curl http://localhost");
```
Comment (grisu48) [inline tests/containers/podman_bci_systemd.pm:43]:
```suggestion
    validate_script_output("curl http://localhost", qr/testpage123-content/);
```

**Comment by grisu48** [comment]
> Isn't this already tested in BCI test framework?

Yes for BCI images. The purpose is to extend the container engine test to include systemd handling and close a coverage whole in the engine test runs.

File: tests/containers/podman_bci_systemd.pm (hunk 2 above)
Comment (ilausuch) [reply tests/containers/podman_bci_systemd.pm:12]: oh yes, sorry!

File: tests/containers/podman_bci_systemd.pm [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    $self->{podman} = $podman;
+
+    my $image = get_var("CONTAINER_IMAGE_TO_TEST", "registry.suse.com/bci/bci-init:latest");
+
+    record_info('Test', 'Launch a container with systemd');
+    assert_script_run("podman run -d -p 80:80 --name nginx $image");
+
+    record_info('Test', 'Install nginx');
+    assert_script_run("podman exec -t nginx zypper -n in nginx");
+    assert_script_run("podman exec -t nginx bash -c 'echo testpage123-content > /srv/www/htdocs/index.html'");
```
Comment (ricardobranco777) [inline tests/containers/podman_bci_systemd.pm:28]:
```suggestion
    assert_script_run("podman exec nginx zypper -n in nginx");
    assert_script_run("podman exec nginx bash -c 'echo testpage123-content > /srv/www/htdocs/index.html'");
```

File: tests/containers/podman_bci_systemd.pm [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+
+    record_info('Test', 'Start nginx');
+    assert_script_run("podman exec -t nginx systemctl start nginx");
```
Comment (ricardobranco777) [inline tests/containers/podman_bci_systemd.pm:31]:
```suggestion
    assert_script_run("podman exec nginx systemctl start nginx");
```

**Review by ricardobranco777 — CHANGES_REQUESTED** [review]

File: tests/containers/podman_bci_systemd.pm (hunk 7 above)
Comment (ilausuch) [reply tests/containers/podman_bci_systemd.pm]: done

File: tests/containers/podman_bci_systemd.pm (hunk 8 above)
Comment (ilausuch) [reply tests/containers/podman_bci_systemd.pm]: done

File: tests/containers/podman_bci_systemd.pm (hunk 9 above)
Comment (ilausuch) [reply tests/containers/podman_bci_systemd.pm]: done

File: tests/containers/podman_bci_systemd.pm (hunk 10 above)
Comment (ilausuch) [reply tests/containers/podman_bci_systemd.pm]: done

File: tests/containers/podman_bci_systemd.pm (hunk 11 above)
Comment (ilausuch) [reply tests/containers/podman_bci_systemd.pm]: done

**Comment by ilausuch** [comment]
@grisu48 @ricardobranco777  thanks for the reviews

**Review by ricardobranco777 — APPROVED** [review]
LGTM.  A new VR is needed I guess.

**Comment by ilausuch** [comment]
The VR is for the last commit :)

File: tests/containers/podman_bci_systemd.pm [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+    record_info('Test', 'Curl localhost from container');
+    validate_script_output("podman exec nginx curl -sfL http://localhost", qr/testpage123-content/);
+
+    record_info('Test', 'Curl localhost from host');
+    validate_script_output("curl http://localhost", qr/testpage123-content/);
+}
+
+sub cleanup {
+    my ($self) = @_;
+    $self->{podman}->cleanup_system_host();
```
Comment (grisu48) [inline tests/containers/podman_bci_systemd.pm:44]:
```suggestion
    script_run("podman stop nginx");
    $self->{podman}->cleanup_system_host();
```

Comment (ilausuch) [reply tests/containers/podman_bci_systemd.pm:44]: ok

**Comment by ilausuch** [comment]
The VR passes the test with the last change

**Comment by m-dati** [comment]
Approved :+1:

**Comment by jlausuch** [comment]
This is failing in SLE Micro https://openqa.suse.de/tests/12141713#step/podman_bci_systemd/39 and it will fail in products under development wich register against SCC Proxy.
You need an additional step to disable `container-suseconnect-zypp` so that repos from the host don't get propagated with the wrong URL to the container. nginx should be available in the BCI repo itself.

**Comment by DimStar77** [comment]
Multiple fails seen on TW:

https://openqa.opensuse.org/tests/3580555#step/podman_bci_systemd/39

**Comment by jlausuch** [comment]
> Multiple fails seen on TW:
> 
> https://openqa.opensuse.org/tests/3580555#step/podman_bci_systemd/39

The problem in TW is that `CONTAINER_IMAGE_TO_TEST` is used instead of the init image. I'll fix that.

**Pull request merged**